pyflakes/checker.py

```diff
@@ -952,21 +968,27 @@ def IMPORT(self, node):
 
     def IMPORTFROM(self, node):
         if node.module == '__future__':
-            if not self.futuresAllowed:
+            # __future__ can only appear in module/doctest scope and
+            # should not have been disabled already in handleNode and
+            # the scope must not have any other type of binding.
+            if (not isinstance(self.scope, ModuleScope) or
+                    not self.scope._futures_allowed or
+                    any(not isinstance(binding, FutureImportation)
+                        for binding in self.scope.values())):
+                self.scope._futures_allowed = False
```

Comment:
The check for "any other binding in scope" because this method currently can not follow the old pattern

```
if node.module == '__future__':
   ...
else:
   self.futuresAllowed = False
```

"self.futuresAllowed = False" can not become "self.scope._futures_allowed = False" , because the current scope may not be a ModuleScope/DocScope, so "self.scope._futures_allowed" would be an attribute error.

We could work around that by using a more complex else clause like

```
if node.module == '__future__':
   ...
elif hasattr(self.scope, '_futures_allowed'):
   self.scope._futures_allowed = False
```

However that adds an extra branch to every 'import from' processed, whereas a "any other binding in scope" inside `if node.module == '__future__'` achieves the same result and is only executed within the specific `__future__` import nodes, so it will only negatively impact performance when a module has many `__future__` imports.